src/H5FDros3.c

```diff
@@ -599,6 +603,236 @@ H5FD__ros3_fapl_free(void *_fa)
+
+    FUNC_ENTER_API(FAIL)
+    H5TRACE2("e", "i*s", fapl_id, token);
+
+#if ROS3_DEBUG
+    HDfprintf(stdout, "H5Pget_fapl_ros3_token() called.\n");
+#endif
+
+    if (token_dst == NULL)
+        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "token_dst is NULL")
+
+    if (NULL == (plist = H5P_object_verify(fapl_id, H5P_FILE_ACCESS)))
+        HGOTO_ERROR(H5E_PLIST, H5E_BADTYPE, FAIL, "not a file access property list")
+    if (H5FD_ROS3 != H5P_peek_driver(plist))
+        HGOTO_ERROR(H5E_PLIST, H5E_BADVALUE, FAIL, "incorrect VFL driver")
+    if ((token_exists = H5P_exist_plist(plist, ROS3_TOKEN_PROP_NAME)) < 0)
+        HGOTO_ERROR(H5E_PLIST, H5E_CANTGET, FAIL, "failed to check if property token exists in plist")
+    if (token_exists) {
+        if (H5P_get(plist, ROS3_TOKEN_PROP_NAME, token_src) < 0)
+            HGOTO_ERROR(H5E_PLIST, H5E_CANTGET, FAIL, "unable to get token value")
+    }
+
+    /* Copy the token data out */
+    tokenlen = HDstrlen(token_src);
+    if (size < tokenlen) {
+        tokenlen = size - 1;
+    }
+    H5MM_memcpy(token_dst, token_src, sizeof(char) * tokenlen);
+    token_dst[tokenlen] = '\0';
+
+done:
+    FUNC_LEAVE_API(ret_value)
+} /* end H5Pget_fapl_ros3_token() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5FD__ros3_str_token_copy()
+ *
+ * Purpose:     Create a copy of the token string.
+ *
+ * Return:      SUCCEED/FAIL
+ *
+ * Programmer:  Jan-Willem Blokland
+ *              2023-05-26
+ *
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+H5FD__ros3_str_token_copy(const char *name, size_t size, void *_value)
+{
+    char **value     = (char **)&_value;
+    herr_t ret_value = SUCCEED;
+
+    FUNC_ENTER_PACKAGE
+
+#if ROS3_DEBUG
+    HDfprintf(stdout, "H5FD__ros3_str_token_copy() called.\n");
+#endif
+
+    if (*value)
+        if (NULL == (*value = HDstrdup(*value)))
+            HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL, "can't copy string property token")
+
+done:
+    FUNC_LEAVE_NOAPI(ret_value);
+} /* H5FD__ros3_str_token_copy() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5FD__ros3_str_token_cmp()
+ *
+ * Purpose:     Compares two token strings with each other.
+ *
+ * Return:      SUCCEED/FAIL
+ *
+ * Programmer:  Jan-Willem Blokland
+ *              2023-05-26
+ *
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+H5FD__ros3_str_token_cmp(const void *_value1, const void *_value2, size_t size)
+{
+    char const *value1    = (char const *)_value1;
+    char const *value2    = (char const *)_value2;
+    int         ret_value = SUCCEED;
+
+    FUNC_ENTER_PACKAGE_NOERR
+
+    if (value1) {
+        if (value2)
+            ret_value = HDstrcmp(value1, value2);
+        else
+            ret_value = 1;
+    }
+    else {
+        if (value2)
+            ret_value = -1;
```

Comment:
The `FAIL`/`SUCCEED` macros would be more clear.

Reply:
It turns out I made a mistake in the function declaration the return type should have been `int` instead of `herr_t`. I will fix it. This means that your suggestion to make use of `FAIL`/`SUCCEED` marco make less sense now.